### PR TITLE
Benchmarking partitioning

### PR DESCRIPTION
**Pull request by knighton**
## Description of changes:

Benchmarking partitioning.

## Merge Checklist:
_Put an `x` without space in the boxes that apply. If you are unsure about any checklist, please don't hesitate to ask. We are here to help! This is simply a reminder of what we are going to look for before merging your pull request._

### General
- [x] I have read the [contributor guidelines](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md)
- [ ] This is a documentation change or typo fix. If so, skip the rest of this checklist.
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the MosaicML team.
- [ ] I have updated any necessary documentation, including [README](https://github.com/mosaicml/streaming/blob/main/README.md) and [API docs](https://github.com/mosaicml/streaming/tree/main/docs) (if appropriate).

### Tests
- [x] I ran `pre-commit` on my change. (check out the `pre-commit` section of [prerequisites](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#prerequisites))
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate).
- [x] I ran the tests locally to make sure it pass. (check out [testing](https://github.com/mosaicml/streaming/blob/main/CONTRIBUTING.md#submitting-a-contribution))
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes.

### CODE REVIEW

**Comment by knighton** [comment]
partition

**Comment by knighton** [comment]
shuffle

File: scripts/partition/bench.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    args.add_argument(
+        '--timeout',
+        type=float,
+        default=120,
+        help='Maximum time to wait for partitioning to complete',
+    )
+    return args.parse_args()
+
+
+def main(args: Namespace) -> None:
```
Comment (karan6181) [inline scripts/partition/bench.py:82]: Does this script saves Partition benchmarking results to a file ?

File: scripts/partition/plot.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+
+def parse_args() -> Namespace:
+    """Parse command-line arguments.
+
+    Returns:
+        Namespace: Command-line arguments.
+    """
+    args = ArgumentParser()
+    args.add_argument('--in', type=str, required=True, help='Partition benchmarking results')
```
Comment (karan6181) [inline scripts/partition/plot.py:36]: Who(or which script) writes the benchmarking results ?

File: scripts/partition/plot.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+from argparse import ArgumentParser, Namespace
+from typing import Dict, List, Tuple
+
+from matplotlib import pyplot as plt
+
+mark_sizes = [
+    ('CIFAR-10 (50K)', 50_000, '#dfd'),
+    ('ImageNet (1.2M)', 1_281_167, '#cfc'),
+    ('C4 (365M)', 364_868_892, '#8c8'),
+    ('LAION-5B (5.85B)', 5_850_000_000, '#484'),
```
Comment (karan6181) [inline scripts/partition/plot.py:15]: Can this be user args ?

File: scripts/shuffle/bench.py [hunk 4]
```diff
@@ -102,8 +134,8 @@ def main(args: Namespace) -> None:
     Args:
         args (Namespace): Command-line arguments.
     """
-    names = 'py1x', 'py2x'
-    get_shuffles = get_shuffle_py1s, get_shuffle_py2s
+    names = 'naive', 'py2s', 'py1s', 'py1b'
```
Comment (karan6181) [inline scripts/shuffle/bench.py:137]: Can this be user provided args ?

File: scripts/shuffle/plot.py [hunk 5]
```diff
@@ -9,9 +9,10 @@
 from matplotlib import pyplot as plt
 
 mark_sizes = [
+    ('CIFAR-10 (50K)', 50_000, '#dfd'),
     ('ImageNet (1.2M)', 1_281_167, '#cfc'),
     ('C4 (365M)', 364_868_892, '#8c8'),
-    ('LAION-5B (5B)', 5_000_000_000, '#484'),
+    ('LAION-5B (5.85B)', 5_850_000_000, '#484'),
```
Comment (karan6181) [inline scripts/shuffle/plot.py:15]: Can this be user provided args ?

**Review by karan6181 — COMMENTED** [review]
Looks good overall. Can we have a readme of under each subdirectory on 
- how to run the scripts ?
- what are the different ways to run the scripts ?
- How to interpret the results
- What user would get out of it after running the script ? 

Have the above bullet points answered in a doc would be very helpful to any community users

File: scripts/partition/bench.py (hunk 1 above)
Comment (snarayan21) [reply scripts/partition/bench.py:82]: Also, we don't currently have another partitioning algo other than `orig` right? just wanted to check

File: scripts/partition/plot.py (hunk 2 above)
Comment (snarayan21) [reply scripts/partition/plot.py:36]: also just curious about the PR overall -- this is to let users benchmark how long different partitions/shuffles take with different configurations of NCN, physical nodes, ranks, workers, batch size, etc. right? as well as see how this performance varies across some typical datasets?

File: scripts/partition/bench.py (hunk 1 above)
Comment (knighton) [reply scripts/partition/bench.py:82]: Yes, we write the results to /dev/stdout in this same method for maximum compatibility with Unix tooling

Comment (knighton) [reply scripts/partition/bench.py:82]: We (james) have experimented with alternative partitioning algorithms in the past, for which this code was originally written, hence this works across algos

File: scripts/partition/plot.py (hunk 3 above)
Comment (knighton) [reply scripts/partition/plot.py:15]: I looked into packing this data into a string flag, and am thinking no, this should remain as is

- There is no use case where you would need to dynamically vary these dataset size/time comparison lines
- Horizontally, if the dataset size comparisons could be improved, you'd want to fix it permanently for all plots by editing the source, not temporarily by editing a flag
- Vertically, if the arrangement of "Instant", "Fast", and "Slow" could be improved, you'd want to fix it permanently for all plots by editing the source, not temporarily by editing a flag
- Flags values like `CIFAR-10 (50K),50000,#dfd;ImageNet (1.2M),1281167,#cfc;C4 (365M),364868892,#8c8;LAION-5B (5.85B),5850000000,#484` are not human-friendly
- Flag packing/unpacking would be half the file while not improving anything, etc

File: scripts/partition/plot.py (hunk 2 above)
Comment (knighton) [reply scripts/partition/plot.py:36]: For both partitioning and shuffling, there is a `bench.py` whose benchmarking results are piped to a file which is then plotted by `plot.py`

Comment (knighton) [reply scripts/partition/plot.py:36]: Right, given various factors you've listed, see what the scalability curve for generating the epoch looks like on your specific system, exactly, without having to actually generate many fake datasets of all that range of sizes and train on them.

File: scripts/shuffle/plot.py (hunk 5 above)
Comment (knighton) [reply scripts/shuffle/plot.py:15]: As above

File: scripts/shuffle/bench.py (hunk 4 above)
Comment (knighton) [reply scripts/shuffle/bench.py:143]: Done (added `--algos` flag)

**Review by karan6181 — APPROVED** [review]
Based on the offline discussion, there will be a new PR which adds the readme on how to run the script and how to interpret the results!

**Pull request merged**